File: Dev/unitTest_Engine_cpp_gtest/ParentObject.cpp
#include <ace.h>
#include <gtest/gtest.h>
#include <memory>
#include "EngineTest.h"

using namespace std;
using namespace ace;

class ObjectSystem_ParentObject : public EngineTest
{
public:
	ObjectSystem_ParentObject(bool isOpenGLMode)
		: EngineTest(ace::ToAString("ParentObject"), isOpenGLMode, 120)
		, m_parent(nullptr)
	{
	}

protected:
	shared_ptr<TextureObject2D> m_parent;

	void OnStart()
	{
		auto scene = make_shared<Scene>();
		auto layer = make_shared<Layer2D>();
		auto child = make_shared<TextureObject2D>();
		m_parent = make_shared<TextureObject2D>();

		GetEngine()->ChangeScene(scene);
		scene->AddLayer(layer);
		layer->AddObject(m_parent);
		layer->AddObject(child);
		m_parent->AddChild(child, eChildMode::CHILD_MODE_ALL);

		auto g = ace::GetGraphics();
		auto texture = g->CreateTexture2D(ace::ToAString("Data/Texture/Cloud1.png").c_str());
		m_parent->SetTexture(texture);
		child->SetTexture(texture);

		m_parent->SetPosition(Vector2DF(320, 240));
		m_parent->SetCenterPosition(Vector2DF(128, 128));
		m_parent->SetColor(Color(255, 255, 128, 255));
		child->SetPosition(Vector2DF(50, 50));
	}

	void OnUpdating()
	{
		m_parent->SetAngle(m_parent->GetAngle() + 0.2f);
		m_parent->SetScale(m_parent->GetScale() + Vector2DF(0.001f, 0.001f));
	}
};

void Test_ParentObject()
{
	RunTest<ObjectSystem_ParentObject>(true);
	RunTest<ObjectSystem_ParentObject>(false);
}

TEST(ObjectSystem, ParentObject_GL)
{
	RunTest<ObjectSystem_ParentObject>(true);
}

#if _WIN32
TEST(ObjectSystem, ParentObject_DX)
{
	RunTest<ObjectSystem_ParentObject>(false);
}
#endif

File: Dev/test_cs/ObjectSystem/ParentObject.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using ace;

namespace test_cs.ObjectSystem
{
	class ParentObject : ISample
	{
		public void Run()
		{
			Engine.Initialize("ParentObject", 1280, 720, new EngineOption());

			var scene = new Scene();
			var layer = new Layer2D();
			var parent = new TextureObject2D();
			var child = new TextureObject2D();

			parent.AddChild( child, ChildMode.All );
			layer.AddObject( parent );
			layer.AddObject( child );
			scene.AddLayer( layer );
			Engine.ChangeScene( scene );

			var texture = Engine.Graphics.CreateTexture2D("Data/Texture/Cloud1.png");

			parent.Position = new Vector2DF() { X = 640, Y = 360 };
			parent.CenterPosition = new Vector2DF() { X = 128, Y = 128 };
			parent.Texture = texture;
			child.Position = new Vector2DF() { X = 50, Y = 50 };
			child.Texture = texture;

			while( Engine.DoEvents() )
			{
				parent.Angle += 0.1f;
				parent.Scale = new Vector2DF() { X = parent.Scale.X + 0.0005f, Y = parent.Scale.Y + 0.0005f };
				Engine.Update();
			}

			Engine.Terminate();
		}
	}
}
